src/NuGet.Core/NuGet.Packaging/Signing/Authoring/X509SignatureProvider.cs

```diff
@@ -61,43 +62,83 @@ public Task<Signature> CreateSignatureAsync(SignPackageRequest request, Signatur
 #if IS_DESKTOP
         private Signature CreateSignature(SignPackageRequest request, SignatureContent signatureContent)
         {
+            var cmsSigner = CreateCmsSigner(request);
+
+            if (request.PrivateKey != null)
+            {
+                return CreateSignature(cmsSigner, signatureContent, request.PrivateKey);
+            }
+
+            var contentInfo = new ContentInfo(signatureContent.GetBytes());
+            var cms = new SignedCms(contentInfo);
+
+            cms.ComputeSignature(cmsSigner);
+
+            return Signature.Load(cms);
+        }
+
+        private static CmsSigner CreateCmsSigner(SignPackageRequest request)
+        {
+            // Subject Key Identifier (SKI) is smaller and less prone to accidental matching than issuer and serial
+            // number.  However, to ensure cross-platform verification, SKI should only be used if the certificate
+            // has the SKI extension attribute.
+            CmsSigner signer;
+
+            if (request.Certificate.Extensions[Oids.SubjectKeyIdentifier] == null)
+            {
+                signer = new CmsSigner(SubjectIdentifierType.IssuerAndSerialNumber, request.Certificate);
+            }
+            else
+            {
+                signer = new CmsSigner(SubjectIdentifierType.SubjectKeyIdentifier, request.Certificate);
+            }
+
             var attributes = SigningUtility.GetSignAttributes(request);
 
-            using (request)
+            foreach (var attribute in attributes)
             {
-                if (request.PrivateKey != null)
-                {
-                    return CreateSignature(request.Certificate, signatureContent, request.PrivateKey, request.SignatureHashAlgorithm, attributes, request.AdditionalCertificates);
-                }
+                signer.SignedAttributes.Add(attribute);
+            }
 
-                var contentInfo = new ContentInfo(signatureContent.GetBytes());
-                var cmsSigner = new CmsSigner(SubjectIdentifierType.SubjectKeyIdentifier, request.Certificate);
+            using (var chain = new X509Chain())
```

Comment:
If we are going to build the chain in advance we should also pass the list of certificates down to `SigningUtility.GetSignAttributes` instead of letting it build a chain also for the `signing-certificate-v2` attribute.

Currently the verify uses the same chain as the trust validation, but for signing it wasn't since cms was handling this before